### PR TITLE
sql: rewrite index backfiller to stream back progress

### LINKED ISSUE #54528: sql, bulk: rewrite index backfiller
Currently it does fixed size batches, and also starts/stops regularly for progress reporting.

Instead, we could look more like IMPORT: produce in small batches, stream and flush as needed, return async progress as we go, etc. Sitting on the old backfiller / chunkbacfiller is probably doing more harm than good now and a scratch rewrite of the build index piece could make it smaller, more reliable and faster.

### PR DESCRIPTION
**Pull request by adityamaru**
Previously, the DistSQL version index backfiller did a a few things
which were non-idiomatic:

- There was fixed time duration and a buffer fill threshold at which the
  backfiller would stop constructing/ingesting index entries and update
the job with its progress. This would potentially cause the DistSQL flow
to get planned, run and tore down multiple times within a single index
backfill schema change.

- The index backfiller relied on the chunk backfiller which performed
  all operations on a fixed "chunk" of rows at a time. Future plans to
make this a memory account/dynamically changing unit of operation meant
that we had to move away from this model.

The new index backfiller logic takes inspiration from our import
coordinator and processors setup. In general, the DistSQL flow is setup
once and the various processors are assigned the spans to backfill.
Every processor concurrently constructs one or more "index entry batch"
and ingests the associated KVs. Periodic go routines stream back the
completed spans to the coordinator node, which in turn periodically
writes the fraction progressed and remaining spans (to account for job
resumption) to the schema change job.

While the size of the index entries in a batch is currently fixed,
future work to be memory aware via the index backfiller mem monitor
should not be particularly difficult to add.

It is important to note that the core index entry building logic has not
been changed. Neither has the column backfiller or the logic for index
backfilling in a txn.

Fixes: #54528

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
Marked as WIP because I will still work on improving logging, comments, and progress tests. I will add another commit which deletes the unused old indexbackfiller code path.

There are some TODOs which I would like clarification on as I am not familiar with table descriptor related nuances, I have highlighted these.

After our meeting with the schema team, I realize that the backfiller will undergo more iterations based on where we want to start writing todoSpans/progress etc. Decoupling the index backfiller from the chunk backfiller seemed a good first step regardless. Open to all comments and refactor suggestions!

File: pkg/sql/backfill.go [hunk 1]
```diff
@@ -886,6 +893,275 @@ func (sc *SchemaChanger) nRanges(
...
+	}); err != nil {
+		return err
+	}
+
+	log.VEventf(ctx, 2, "indexbackfill: initial resume spans %+v", todoSpans)
+
+	var p PhysicalPlan
+	var evalCtx extendedEvalContext
+	var planCtx *PlanningCtx
+	// TODO(adityamaru): Is it okay to run the DistSQL flow and progress update
```
Comment (adityamaru) [inline pkg/sql/backfill.go:954]: Calling this out as a decision I took with potentially limited understanding of its implication.

File: pkg/sql/schema_changer_test.go [hunk 2]
```diff
@@ -1356,6 +1364,15 @@ func TestSchemaChangeRetryOnVersionChange(t *testing.T) {
 				if currChunk == 3 {
 					// Publish a new version of the table.
 					upTableVersion()
+
+					// TODO(adityamaru): Previously, the index backfiller would
```
Comment (adityamaru) [inline pkg/sql/schema_changer_test.go:1368]: Is not observing changes in table desc version throughout an index backfill okay? This was not the case in the old implementation as we would re-acquire the table desc lease after every 2mins/80% buffer fill as part of the DistSQL flow planning. This in turn would throw a version mismatch error and our schema change would be retried.

Comment (dt) [reply pkg/sql/schema_changer_test.go:1368]: I'll defer to @lucy-zhang or @ajwerner on this one.

**Comment by adityamaru** [comment]
👋 I want to pick this back up sometime soon, friendly ping for a review!

**Review by pbardea — COMMENTED** [review]
Took a first pass at the current approach. I want to do another pass at a higher level, but thought to get these comments out before I do another pass (I'm still digesting these changes).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-:-MM1S8jA72p0p2Ia0roL:b-8gfln8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, @lucy-zhang, and @pbardea)

___
*[pkg/sql/backfill.go, line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1PnSMCoANQh9g0dDz:-MM1PnSMCoANQh9g0dE-:b1c2vjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L76)):*
> ```Go
> 	// indexBackfillBatchSize is the maximum number of index entries we attempt to
> 	// fill in a single index batch before queueing it up for ingestion and
> 	// progress reporting in the index backfiller processor.
> ```

Do we have commentary anywhere that describes the relationship between these "batches" that are referred to here and the "chunks" referred to above? Or what determines the actual number of index entries that can go into a single batch that we can point to?
___
*[pkg/sql/backfill.go, line 983 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1L1Z4DWBcDgslSHTm:-MM1L1Z4DWBcDgslSHTn:bsbayl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L983)):*
> ```Go
> 	// streaming back progress and the updates to the job details/progress
> 	// fraction.
> 	var mu sync.Mutex
> ```

Instead of maintaining this bare mutex, what do you think of having a struct that looks like:
```
	mu := struct {
		syncutil.Mutex
		updatedTodoSpans []roachpb.Span
	}{}
```
I prefer a struct like that to make any access to `updatedTodoSpans` explicitly need to go through `mu`.
___
*[pkg/sql/backfill.go, line 1036 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1IdCS4095q3idW673:-MM1IdCS4095q3idW674:b-h8jq0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L1036)):*
> ```Go
> 			// No processor has returned completed spans yet.
> 			if updatedTodoSpans == nil {
> 				return nil
> ```

Do we need a `mu.Unlock()` here? 
___
*[pkg/sql/rowexec/indexbackfiller.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1N1y68aN2J3yk57ac:-MM1N1y68aN2J3yk57ad:bgmw3vp) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/rowexec/indexbackfiller.go#L191)):*
> ```Go
> 	// BulkAdder flush, but is read when progress is being sent back to the
> 	// coordinator.
> 	var mu sync.Mutex
> ```

Same comment about creating a wrapping struct that embeds a mutex for `completedSpans`. It also looks like addedSpans should be included too?
___
*[pkg/sql/rowexec/indexbackfiller.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1RMcrC5j7JuriZ2BM:-MM1RMcrC5j7JuriZ2BN:bdy3qsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/rowexec/indexbackfiller.go#L245)):*
> ```Go
> 			// span representing this indexBatch as "added". This span will be part of
> 			// the set of completed spans on the next bulk adder flush.
> 			addedSpans = append(addedSpans, indexBatch.completedSpan)
> ```

Do we need a mutex around this? I'm worried about races with the adder flushing callback.
___
*[pkg/sql/rowexec/indexbackfiller.go, line 472 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1Nd0C6P_wdEMiqmSO:-MM1Nd0C6P_wdEMiqmSP:b-nt7m7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/rowexec/indexbackfiller.go#L472)):*
> ```Go
> 	prepTime := timeutil.Now().Sub(start)
> 	if log.V(3) {
> 		log.Infof(ctx, "index backfill stats: entries %d, prepare %+v",
> ```

Curious if there was any specific reason to go with `log.Infof` over `log.VEventf(3, ...`?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @lucy-zhang, and @pbardea)

___
*[pkg/sql/backfill.go, line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1PnSMCoANQh9g0dDz:-MMbowEg12KIYjXAbXpA:bb73h4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L76)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Do we have commentary anywhere that describes the relationship between these "batches" that are referred to here and the "chunks" referred to above? Or what determines the actual number of index entries that can go into a single batch that we can point to?
</blockquote></details>

Yeah I agree it is a little unfortunate that there exist two units of execution, "chunk" and a "batch" but they're essentially the same thing at the lowest level. It is the number of entries that `BuildIndexEntriesChunk` builds before returning from the method.

I introduced "batch" as a unit of operation in the distributed index backfill (see indexEntryBatch struct). We construct a batch of index entries and stream the batch for ingestion. The ingester, adds all the entries in a batch into the buffering adder and considers those entries as "completed" for the purpose of progress reporting.

The concept of a "chunk" is a remnant prior to this refactor. For the index backfiller it is now only used when the index backfill is run in a txn, to represent how many index entries we will build before we InitPut them. I could just change the txn index backfill to also use "batch" to represent this?

To answer your second question, both the size of a chunk and the size of a batch are currently controlled by hardcoded values. I see a future add on where this is a little more dynamic based on the memory we have.
___
*[pkg/sql/backfill.go, line 983 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1L1Z4DWBcDgslSHTm:-MMbWwYD9C66yOxzI04S:b-21w2pp) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L983)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Instead of maintaining this bare mutex, what do you think of having a struct that looks like:
```
	mu := struct {
		syncutil.Mutex
		updatedTodoSpans []roachpb.Span
	}{}
```
I prefer a struct like that to make any access to `updatedTodoSpans` explicitly need to go through `mu`.
</blockquote></details>

good idea, done.
___
*[pkg/sql/backfill.go, line 1036 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1IdCS4095q3idW673:-MMbWsKdA64nO8gb5aP6:b-aia7ss) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L1036)):*
> Instead of maintaining this bare mutex, what do you think of having a struct that looks like:
ah good catch!
___
*[pkg/sql/rowexec/indexbackfiller.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1N1y68aN2J3yk57ac:-MMb_t8F4La0jqoFVC-m:bl7tnra) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/rowexec/indexbackfiller.go#L191)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Same comment about creating a wrapping struct that embeds a mutex for `completedSpans`. It also looks like addedSpans should be included too?
</blockquote></details>

good point, done.
___
*[pkg/sql/rowexec/indexbackfiller.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1RMcrC5j7JuriZ2BM:-MMb_uIi5N5qbfQb5LN8:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/rowexec/indexbackfiller.go#L245)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Do we need a mutex around this? I'm worried about races with the adder flushing callback.
</blockquote></details>

done.
___
*[pkg/sql/rowexec/indexbackfiller.go, line 472 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MM1Nd0C6P_wdEMiqmSO:-MMbaMJAATeCkQQURWK_:b-ps02ji) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/rowexec/indexbackfiller.go#L472)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Curious if there was any specific reason to go with `log.Infof` over `log.VEventf(3, ...`?
</blockquote></details>

none at all :) changed.

**Comment by adityamaru** [comment]
@pbardea this should be RFAL. Addressed your comments and added some more comments to the code.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @lucy-zhang, and @pbardea)

___
*[pkg/sql/backfill.go, line 954 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MJTqzlVfU58Kmya9Poo-r1-954:-MMbx65_2lFNNpxdWgh0:bkpxfdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L954)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Calling this out as a decision I took with potentially limited understanding of its implication.
</blockquote></details>

@ajwerner i'm increasingly convinced that my use of the leased table desc is incorrect here. I do not fully appreciate the semantics of leases yet but it seems like releasing the lease before the index has been backfilled is incorrect? Should the run method of the distsql flow also be invoked in this txn?

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @lucy-zhang, and @pbardea)

___
*[pkg/sql/backfill.go, line 954 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MJTqzlVfU58Kmya9Poo-r1-954:-MNUrpsM9eUG7DGPJ1DG:b-r8w4zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L954)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

@ajwerner i'm increasingly convinced that my use of the leased table desc is incorrect here. I do not fully appreciate the semantics of leases yet but it seems like releasing the lease before the index has been backfilled is incorrect? Should the run method of the distsql flow also be invoked in this txn?
</blockquote></details>

This is nuanced and unfortunate. There's certainly tradeoffs going on here. The safe thing to do would be to hold onto a lease throughout the duration of the backfill and ensure its validity. Even that, however, is hard as we don't auto-refresh leases and nothing is checking expiration as it stands.

Also, even if we could meaningfully hold on to the lease the entire time, it'd be rather unfortunate. The downside of holding on to that lease is that there are certain schema changes we'd like to permit to finish, such as changing privileges or enqueuing other schema changes. 

We need to think about the set of things which can happen to this descriptor while this backfill is running. Any changes to columns or indexes should be blocked so in principle we should be okay not leasing the descriptor. Relying on this, however, is quite subtle. 

One thing I worry about is what happens upon truncate. @lucy-zhang noted that we don't end up doing anything with those jobs and probably need to.

A more principled solution might be to have the coordinator acquire a fresh lease and then drop the lease between checkpoints but that's awkward in terms of the code and not clear exactly what you'd do when you got the new lease in terms of checking the validity. 

tl;dr I think what you're doing here ends up being okay but only for implicit reasons. Augment the comment to indicate that this index used for planning cannot be invalidated as any schema change which would invalidate it would be queued behind it in the mutations slice.

**Review by pbardea — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, @lucy-zhang, and @pbardea)

___
*[pkg/sql/schema_changer.go, line 1947 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MP0-lFO41uPz_oBzQlF:-MP0-lFO41uPz_oBzQlG:biabqga) ([raw file](https://github.com/cockroachdb/cockroach/blob/6974d8957f022c8ea111135b88e42f98a23d8fff/pkg/sql/schema_changer.go#L1947)):*
> ```Go
> 	AlwaysUpdateIndexBackfillProgress bool
> 
> 	// transaction is unable to commit because it is violating the two
> ```

It looks like the line`// TwoVersionLeaseViolation is called whenever a schema change` was accidentally removed here.

**Review by thoszhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, and @lucy-zhang)

___
*[pkg/sql/backfill.go, line 954 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MJTqzlVfU58Kmya9Poo-r1-954:-MQTdoAq3v3pENT5gxON:bcpfiu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L954)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This is nuanced and unfortunate. There's certainly tradeoffs going on here. The safe thing to do would be to hold onto a lease throughout the duration of the backfill and ensure its validity. Even that, however, is hard as we don't auto-refresh leases and nothing is checking expiration as it stands.

Also, even if we could meaningfully hold on to the lease the entire time, it'd be rather unfortunate. The downside of holding on to that lease is that there are certain schema changes we'd like to permit to finish, such as changing privileges or enqueuing other schema changes. 

We need to think about the set of things which can happen to this descriptor while this backfill is running. Any changes to columns or indexes should be blocked so in principle we should be okay not leasing the descriptor. Relying on this, however, is quite subtle. 

One thing I worry about is what happens upon truncate. @lucy-zhang noted that we don't end up doing anything with those jobs and probably need to.

A more principled solution might be to have the coordinator acquire a fresh lease and then drop the lease between checkpoints but that's awkward in terms of the code and not clear exactly what you'd do when you got the new lease in terms of checking the validity. 

tl;dr I think what you're doing here ends up being okay but only for implicit reasons. Augment the comment to indicate that this index used for planning cannot be invalidated as any schema change which would invalidate it would be queued behind it in the mutations slice. 
</blockquote></details>

The only benefit of using the leased descriptor here is for the caching, right? I agree that we don't get any correctness guarantees from holding the lease.

Also, now that I think about it: What is preventing us from getting a leased descriptor on the previous version v-1 (i.e., with the index being in delete-only, not in delete-and-write-only) here? Do we wait for old leases to drain anywhere between when we write the descriptor at version v and this transaction? Shouldn't leases at version v-1 still be valid here? Maybe it doesn't actually matter whether we get version v-1 or v, since we just need to read the index descriptor (or whatever), and that doesn't change between versions. I think this is a separate concern from the one you're addressing.
___
*[pkg/sql/schema_changer_test.go, line 1368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MJTqzlVfU58Kmya9Pov-r1-1368:-MQ_1v2RD-zRSCVxX_tv:b-5axfk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/schema_changer_test.go#L1368)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'll defer to @lucy-zhang or @ajwerner on this one.
</blockquote></details>

I think the discussion in the other thread should cover the correctness concerns. Separately, though, we should probably add a test for performing (e.g.) `GRANT` or `REVOKE` in the middle of a backfill (which would bump the table descriptor version) and making sure that it still works, if we don't have that already.

**Review by thoszhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @dt)

___
*[pkg/sql/backfill.go, line 954 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MJTqzlVfU58Kmya9Poo-r1-954:-MQ_9dP2C4KZK_T5KgL5:b-4mbrk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/49cb89a5875278cb12f9a11199d3afe64306fba0/pkg/sql/backfill.go#L954)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

The only benefit of using the leased descriptor here is for the caching, right? I agree that we don't get any correctness guarantees from holding the lease.

Also, now that I think about it: What is preventing us from getting a leased descriptor on the previous version v-1 (i.e., with the index being in delete-only, not in delete-and-write-only) here? Do we wait for old leases to drain anywhere between when we write the descriptor at version v and this transaction? Shouldn't leases at version v-1 still be valid here? Maybe it doesn't actually matter whether we get version v-1 or v, since we just need to read the index descriptor (or whatever), and that doesn't change between versions. I think this is a separate concern from the one you're addressing.
</blockquote></details>

Never mind about the second paragraph. We do call `WaitToUpdateLeases` at the end of `RunStateMachineBeforeBackfill`, so we're fine.

**Review by ajwerner — COMMENTED** [review]
This is a big improvement. What is it going to take to get this merged? Ideally sooner rather than later.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-:-MQqlzqsClBB1gt_cdbS:biwa4p2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @dt)

**Comment by adityamaru** [comment]
> This is a big improvement. What is it going to take to get this merged? Ideally sooner rather than later.
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-:-MQqlzqsClBB1gt_cdbS:biwa4p2)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @dt)

I'm going to pick this up right now and get it merged. Apologies for the delay!

**Comment by adityamaru** [comment]
>GRANT or REVOKE in the middle of a backfill (which would bump the table descriptor version) and making sure that it still works

added a test which grants and revokes from a table desc which is being index backfilled.

>Augment the comment to indicate that this index used for planning cannot be invalidated as any schema change which would invalidate it would be queued behind it in the mutations slice.

added some of your points to the comment.

**Review by thoszhang — COMMENTED** [review]
It looks like the tests may have timed out on CI.

> Reviewed 1 of 8 files at r1, 3 of 7 files at r2, 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-:-MQs5TAY15Va4ab2-oYd:b-qklizg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @ajwerner)

___
*[sqltests, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MQs5dgtAobKNSwJ4_eT:-MQs5dgtAobKNSwJ4_eU:bz6r8op) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7d61ef338190d6b0260e29f33d609944ba3312/sqltests#L1)):*
> ```
> Running make with -j16
> ```

Did you mean to commit this file?
___
*[pkg/sql/schema_changer_test.go, line 3289 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MQs5DK838NbIVyj-BZV:-MQs5DK838NbIVyj-BZW:b-1g1agn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7d61ef338190d6b0260e29f33d609944ba3312/pkg/sql/schema_changer_test.go#L3289)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
> 	// Wait until the mutation is queued up.
> 	testutils.SucceedsSoon(t, func() error {
> 		tableDesc := catalogkv.TestingGetTableDescriptor(kvDB, keys.SystemSQLCodec, "t", "test")
> 		if l := len(tableDesc.Mutations); l != 1 {
> 			return errors.Errorf("number of mutations = %d", l)
> 		}
> 		return nil
> 	})
> ```
</details>


I'm confused about why waiting for the mutation is necessary. Shouldn't the mutation always be present by the time we get the notification from `RunBeforeBackfillChunk`?

**Comment by adityamaru** [comment]
I think there's also a data race in the way the memory monitor is being used during ingestion. Looking into this.

**Comment by adityamaru** [comment]
Just spoke to @pbardea, there is a little bit of work to be done to get the memory accounting thread-safe and accurate, working on it now.

**Review by thoszhang — APPROVED** [review]
> Reviewed 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @ajwerner)

___
*[pkg/sql/schema_changer_test.go, line 3302 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MQtH1eC9TNtFK9gdIKS:-MQtH1eC9TNtFK9gdIKT:bt3o8r4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4ceed6df23df3696916dc05c543885777b69c34/pkg/sql/schema_changer_test.go#L3302)):*
> ```Go
> 
> 	close(continueSchemaChangeNotification)
> 	wg.Wait()
> ```

Could we check again at the end that the grants are still what we expect?

**Review by pbardea — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55468)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, and @lucy-zhang)

___
*[pkg/sql/backfill/backfill.go, line 656 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/55468#-MQwF7MN5xUt0sBrrMJy:-MQwF7MN5xUt0sBrrMJz:b-v44iyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1485409d0a093cccbd725b94e4bab3fe95839ddc/pkg/sql/backfill/backfill.go#L656)):*
> ```Go
> // The method accounts for the memory used by the index entries for this chunk
> // using the memory monitor associated with ib. It is the callers responsibility
> // to clear the associated bound account when appropriate.
> ```

Let's update this comment indicating that it will also return the amount of memory that needs to be freed from the memory monitor when `entries` is released.

**Comment by adityamaru** [comment]
TFTRs!!

bors r+

**Comment by adityamaru** [comment]
looks like a flake:

bors r+
